### ISSUE TITLE
Apache Parquet Support

### ISSUE THREAD

**Issue by bmcdonald3** [Bears-R-Us/arkouda#903]
As I am starting to work on implementing reading in of Parquet files to Chapel arrays, I thought I would summarize what I have been finding here to get feedback and make sure I am going down the right path.

## Summary of Parquet I/O using GLib Interface
- Parquet files store columns of data, each column being of a single data type (although, different columns in the same file could be different data types)
- Columns are stored in "chunk arrays", which are not guaranteed to be contiguous in memory (which, among other reasons, make a simple pointer swap on the buffer difficult)
- Reading in Parquet columns (using the GLib interface) requires that the columns be read into Arrow arrays first, before being copied to Chapel arrays

## Interacting with Parquet files
From what I have gathered, the industry norm for interacting with Parquet files looks something like the following: 
1. Loading in the "table" (that is, all of the columns in the file) into memory
2. Reading in only a specific column on an as-needed basis to interact with
3. Keeping the table in memory for easy-access to other columns when needed

### First proposal: reading specific columns from a file
Based on this information, I have been taking the approach of allowing the reading of specific columns from the Parquet file, which looks something like this:
```chapel
var chplIntArr = getIntColumn(path="./data/file.parquet", col=0);
var chplStrArr = getStringColumn(path="./data/file.parquet", col=1);
```

### Second proposal: reading entire files at a time
```chapel
// tuple size matches the number of columns in the parquet file
(chplIntArr, chplStrArr) = readParquetFile(path="./data/file.parquet", numCols=2);
```

edit: adding this third proposal
### Third proposal: reading columns from in-memory table (not file)
When Arrow files are read, they are stored in tables and since the implementation for how this is done is largely out of our control (and based on the info in the Interacting with Parquet Files section above), this seems to be the most performant method for large files. The table is read in once and then passed to the various methods, with the data already in memory.
```chapel
var table = readParquetFile(path="./data/file.parquet");
var chplIntArr = getIntColumn(table=table, col=0);
var chplStrArr = getStringColumn(table=table, col=1);
```

## Remaining questions
1. Are you expecting multiple columns per file or just a single column per file?
2. Would it be valuable to allow the reading in of Parquet metadata/schema information to Chapel?
3. Is the entire file approach or the column-by-column approach preferred?
4. Does the information here align with your understanding?/Do you have any suggestions or alternate interface proposals?
5. Are the types of the columns known in advance or would it be valuable to have auto-type detection? (e.g, just `getColumn()` rather than `getIntColumn()`

Lastly, any sample files that match what the expected size and format of the Parquet files that you intend to work with would be greatly appreciated.

**Comment by bmcdonald3** [comment]
Thanks for the feedback guys, I have been working on getting everything in good shape and will hopefully have something to show for it soon. One question, however, that hadn't occurred to me when creating this issue is: Is the ability to write to a Parquet file also valuable/important? 

Is the Parquet use-case entirely going to be based on reading in Parquet files and any writing would be done to an HDF5 file and not a Parquet file? If writing to Parquet files is important, what functionality would you envision being useful for that? Thank you.

**Comment by hokiegeek2** [comment]
@bmcdonald3 yes, writing to parquet very important! Basically however we read in, be able to write out.

**Comment by reuster986** [comment]
@bmcdonald3 I second @hokiegeek2 that being able to round-trip arrays to/from parquet is important. One place where this shows up is in the unit tests/CI, where we will typically generate a few random arrays in arkouda, write them to disk, and read them back in to make sure the result is the same as the original. Having that capability with parquet will be very important for writing useful tests (as well as for users to get data out).

**Comment by bmcdonald3** [comment]
Here is a quick summary on the progress that has been made with Parquet so far. 

### Summary 
Using the Arrow GLib C interface, we have been able to get Parquet I/O working in Arkouda, but it is by no means perfect just yet and there is quite a bit of remaining work. Having a discussion at next weeks Arkouda call (or another time) about what is needed on the Arkouda end to get this to a usable place would be valuable for determining where to focus our efforts.

#### read from parquet file
```
>>> ak.read_parquet(['test1.parquet', 'test2.parquet'])
array([0 1 2 3 4 5 6 7 0 1 2 3 4])
```

#### write to parquet file (on 2 locales)
```
>>> a = ak.read_parquet(['test1.parquet', 'test2.parquet'])
>>> a.save_parquet('write-test')
'wrote array to file'
>>> glob.glob('*.parquet')
['write-test_LOCALE0.parquet', 'write-test_LOCALE1.parquet']
>>> ak.read_parquet('write-test_LOCALE0.parquet')
array([0 1 2 3 4 5 6])
```

### Challenges
- Arrow GLib C interface doesn't seem to allow reading of column sizes and other information from metadata (even though I know it is stored)
  - We are experimenting with working with the C++ interface for this task specifically
- The Arkouda HDF5 string array reading doesn't seem to have a parallel with Parquet's API, so that isn't working just yet and will likely need to be done differently than how HDF5 reads string arrays in Arkouda today
- Getting Arrow/Parquet module into Chapel `main` after the release

### Conclusion
In conclusion, I mainly am wondering: 
1. What would need to get done on this effort for it be usable/production ready for Arkouda?
2. Is there any functionality beyond the basic reading/writing shown above you are hoping to see?

I am planning to give a quick demo of the code at the next Arkouda call and would appreciate any feedback there (tagging @glitch from Mike's recommendation).
Here is the branch that I am working on here if anyone is interested: https://github.com/bmcdonald3/arkouda/tree/parquet.

**Comment by hokiegeek2** [comment]
@bmcdonald3 this is really cool, really exciting! Are you also planning on adding read-write capability to the [Plasma Object Store](https://arrow.apache.org/docs/python/plasma.html)?

**Comment by bmcdonald3** [comment]
@hokiegeek2 that has not been a priority of mine as of late, where I have been just focusing on Parquet I/O in regards to Arrow, but I did spend some time looking into the Plasma Object Store prior to this work. It seemed to me there were some pretty strong limitations with it (objects are read-only, must be shared memory, etc.), so we put that on hold for the time being, but we could invest more into that effort if it would be valuable/helpful to the Arkouda team. Is that something that would be valuable to you/others? If so, could you provide an example use-case that you would like to be able to use in Arkouda?

**Comment by hokiegeek2** [comment]
@bmcdonald3 My thoughts are that enabling Chapel I/O with the Plasma Object Store could enable zero-copy data sharing between Chapel locales and distributed Python frameworks such as Ray, Dask, and PySpark as well as distributed Java/Scala frameworks such as Spark. The main use case would be for an Arkouda user to downselect an Arkouda array and then do a zero-copy read of that array into one of these other tools to leverage the Python ML/DL ecosystem (Ray, Dask, PySpark), Spark to enable integration with Spark tools and frameworks, and basically any other distributed framework that utilizes the Plasma Object Store for in-memory data sharing.

It's definitely not an immediate need as this is not an Arkouda development priority. This is an idea I've had for awhile and my plan is to investigate this on my own time to see what's possible. I've just recently started looking into how to make this happen on the Ray/Dask/PySpark/Spark side. The Ray option seems to be the most accessible one as Ray distributes a plasma object store with each Ray worker.

**Comment by bmcdonald3** [comment]
Here is a quick update of the performance that has been observed reading in Parquet files to Chapel arrays with various configurations collected on a linux server:

Config | Read Speeds
-- | --
C/GLib | 0.000344 GiB/sec
C++ RG=1 | 0.000681 GiB/sec
C++ RG=1MB | 0.70564 GiB/sec
C++ RG=256MB | 0.5691 GiB/sec
C++ RG=512MB | 0.6429 GiB/sec
HDF5 | 2.23 GiB/sec

`RG` in this table stands for the size of the "row group" used when writing the file. This is a property that is set by the file writer upon file creation, so it may be outside of the user's control, but how the file was written can significantly alter the performance of reading a Parquet file.

For context, from the Apache Arrow documentation: "A Row Group is a logical horizontal partitioning of the data into rows."

In regards to file writing performance, this is about 15x slower than HDF5, but no work has been done to investigate how we could improve that performance, so I didn't think it was a fair comparison to have it displayed in a table like the read speeds.
